wsapi/wsapiV2.go

```diff
@@ -722,40 +720,55 @@ func HandleV2RawData(state interfaces.IState, params interface{}) (interface{},
 		return nil, NewInvalidHashError()
 	}
 
+	dbase := state.GetDB()
 	var block interfaces.BinaryMarshallable
-	var b []byte
-
-	if block, _ = state.FetchECTransactionByHash(h); block != nil {
-		b, _ = block.MarshalBinary()
-	} else if block, _ = state.FetchFactoidTransactionByHash(h); block != nil {
-		b, _ = block.MarshalBinary()
-	} else if block, _ = state.FetchEntryByHash(h); block != nil {
-		b, _ = block.MarshalBinary()
-	}
-
-	if b == nil {
-		dbase := state.GetDB()
-
-		// try to find the block data in db and return the first one found
+	switch hashkey.Type {
+	case HashTypeFCTTx:
+		block, _ = state.FetchFactoidTransactionByHash(h)
+	case HashTypeECTx:
+		block, _ = state.FetchECTransactionByHash(h)
+	case HashTypeEntry:
+		block, _ = dbase.FetchEntry(h)
+	case HashTypeEBlock:
+		block, _ = dbase.FetchEBlock(h)
+	case HashTypeECBlock:
+		block, _ = dbase.FetchECBlock(h)
+	case HashTypeFBlock:
+		block, _ = dbase.FetchFBlock(h)
+	case HashTypeABlock:
+		block, _ = dbase.FetchABlock(h)
+	case HashTypeDBlock:
+		block, _ = dbase.FetchDBlock(h)
+	default:
+		if block, _ = state.FetchFactoidTransactionByHash(h); block != nil {
+			break
+		}
+		if block, _ = state.FetchECTransactionByHash(h); block != nil {
+			break
+		}
+		if block, _ = dbase.FetchEntry(h); block != nil {
```

Comment:
Same as the other comment, this should be the state lookup function.

Reply:
Does it need both?

Reply:
I think so. Writing entries to the database and the processlist are asynchronous, so you might run into race conditions where you have knowledge of the hash but it's not written to the database yet. I'm not 100% certain it's possible, you'd have to do a deep dive of factomd functionality to double check.

Reply:
doesn't hurt to use both I suppose. I guess both should be used under the HashTypeEntry case as well then...

Reply:
but in another comment you said that state.FetchEntryByHash calls dbase.FetchEntry()

Reply:
> but in another comment you said that state.FetchEntryByHash calls dbase.FetchEntry()

Yes, it's the third and last place to look for an entry. It tries to get it from the processlist and holding map first.